flink-yarn/src/main/java/org/apache/flink/yarn/security/HadoopDelegationTokenProvider.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.yarn.security;
+
+import org.apache.flink.configuration.Configuration;
+
+import org.apache.hadoop.security.Credentials;
+
+import java.util.Optional;
+
+/** Hadoop delegation token provider. */
+public interface HadoopDelegationTokenProvider {
+
+    /** Name of the service to provide delegation tokens. This name should be unique. */
+    String serviceName();
+
+    /**
+     * Return true if delegation tokens are required for this service.
+     *
+     * @param flinkConf Flink configuration
+     * @param hadoopConf Hadoop configuration
+     * @return true if delegation tokens are required
+     */
+    boolean delegationTokensRequired(
```

Comment:
Both `delegationTokensRequired` and `obtainDelegationTokens` take some configuration parameters. I wonder what's the relationship between these configurations? Should `obtainDelegationTokens` check whether DT is required before actually get the DT? If the configurations are identical, maybe we can have an `init` method to configure the provider and other methods won't accept configurations parameters?

Reply:
The `flinkConf` contains the Flink configuration, user could specify some additional configurations when implementing the `HadoopDelegationTokenProvider`. The `hadoopConf` is the hadoop conf for the Yarn cluster the job will be submitted to.
As far as I know, DelegationTokenIssuer will automatically handle the existing tokens case:  
https://github.com/apache/hadoop/blob/19ae0faacc5b1e8ce05f11c555922c0c025dcf3b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/DelegationTokenIssuer.java#L58

The `init()` method is a good idea, I will change to that.

Reply:
I mean it's not clear whether `delegationTokensRequired` and `obtainDelegationTokens` get the same configurations. And what `obtainDelegationTokens` should do if DT is not required at all. We need to clarify these in the API contract.